lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -233,19 +234,31 @@ instance Functor (NextBlocksResult target) where
         RollForward cur bh bs -> RollForward cur bh (fmap f bs)
         RollBackward cur -> RollBackward cur
 
+-- | @FollowAction@ enables the callback of @follow@ to signal if the
+-- chain-following should @Quit@, @Continue@, or of the current callback should
+-- be forgotten and retried (@RetryCurrentStep@).
+--
+-- NOTE: @RetryCurrentStep@ is needed to handle data-races in
+-- 'Cardano.Pool.Metrics'. Aside for that it is arguably a bit out-of-place.
+data FollowAction err
+    = ExitWith err
+    | Continue
+    | RetryCurrentStep
+    deriving (Eq, Show)
+
 -- | Subscribe to a blockchain and get called with new block (in order)!
 follow
-    :: forall target block e0 e1. (Show e0, Show e1)
+    :: forall target block e. (Show e)
     => NetworkLayer IO target block
     -- ^ The @NetworkLayer@ used to poll for new blocks.
     -> Trace IO Text
     -- ^ Logger trace
     -> [BlockHeader]
     -- ^ A list of known tips to start from. Blocks /after/ the tip will be yielded.
-    -> (NE.NonEmpty block -> BlockHeader -> ExceptT e0 IO ())
+    -> (NE.NonEmpty block -> BlockHeader -> IO (FollowAction e))
     -- ^ Callback with blocks and the current tip of the /node/.
     -- @follow@ stops polling and terminates if the callback errors.
-    -> (SlotId -> ExceptT e1 IO ())
+    -> (SlotId -> IO (FollowAction e))
```

Comment:
These were `e0` and `e1` on purpose actually. The errors returned by these two callbacks might be different and we can't just force them to be of the same type.

Reply:
🤔 Sure. I don't think that there is an actual reason in practice though.

Btw, on the side I'm kind of wondering wouldn't be nicer to combine `rollback` and `yield` into one function.